### ISSUE TITLE
POST with a body greater than 1024 chars fail on linux

### ISSUE THREAD

**Issue by JonCubed** [dotnet/corefx#19231]
While running .net core in AWS Lambda, a POST with the HttpClient fails always to certain third-party servers when the body has more than 1024 chars. The following is the exception I get

```javascript
{
    "Message": "An error occurred while sending the request.",
    "Data": {},
    "InnerException": {
        "Message": "Server returned nothing (no headers, no data)",
        "Data": {},
        "InnerException": null,
        "StackTrace": "   at System.Net.Http.CurlHandler.ThrowIfCURLEError(CURLcode error)\n   at System.Net.Http.CurlHandler.MultiAgent.FinishRequest(StrongToWeakReference`1 easyWrapper, CURLcode messageResult)",
        "HelpLink": null,
        "Source": "System.Net.Http",
        "HResult": 52
    },
    "StackTrace": "   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\n   at System.Net.Http.HttpClient.<FinishSendAsync>d__58.MoveNext()\n--- End of stack trace from previous location where exception was thrown ---\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\n   at TestHttpRequestsFunction.Function.<CallRemoteUrl>d__4.MoveNext()",
    "HelpLink": null,
    "Source": "System.Private.CoreLib",
    "HResult": 52
}
```

I've tested the exact same request on windows in .net core and postman and it succeeds. 

I use the following code to reproduce the issue.

```csharp
 var httpClientHandler = new HttpClientHandler { AllowAutoRedirect = false};
 _httpClient = new HttpClient(httpClientHandler, true);

var requestBody = $"{{\r\n  \"testData\": \"{string.Empty.PadRight(999)}\"\r\n    }}",
var content = new StringContent(requestBody, Encoding.UTF8, "application/json");
                
var endpoint = ""; // I can provide a sample server privately
var postResponse = await _httpClient.PostAsync(endpointUrl, content).ConfigureAwait(false);
```
my project.json, nothing special just dependency for AWS Lambda

```javascript
{
    "version": "1.0.0-*",
    "buildOptions": {
    },

    "dependencies": {
        "Microsoft.NETCore.App": {
            "type": "platform",
            "version": "1.1.0"
        },
        "Amazon.Lambda.Core": "1.0.0*",
        "Amazon.Lambda.Serialization.Json": "1.1.0"
    },
    
    "tools": {
        "Amazon.Lambda.Tools": "1.5.0"
    },

    "frameworks": {
        "netcoreapp1.1": {
            "imports": "dnxcore50"
        }
    }
}
```
A little searching pointed to the way that cURL handles requests with a body greater than 1024 chars by adding the `Expect: 100 Continue` header. I tried `_httpClient.DefaultRequestHeaders.ExpectContinue = false;` but the issue persists.

I'm not sure where to go from here as root cause seems to be a low level in the framework, I believe this is mostly likely just a configuration issue and any guidance would be appreciated.

**Comment by stephentoub** [comment]
Thanks, @JonCubed. Can you email me an endpoint I can try?

**Comment by JonCubed** [comment]
@stephentoub emailed

**Comment by karelz** [comment]
Moving to 2.0. @stephentoub please let me know if you disagree / if there is evidence it is not critical.

**Comment by stephentoub** [comment]
@JonCubed, thanks for emailing me the endpoint.  I just tried it, and with netcoreapp1.1, I'm able to repro the behavior you see at first, but then when I set:
```C#
httpClient.DefaultRequestHeaders.ExpectContinue = false;
```
it runs successfully, I see that in a trace that it's successfully suppressing the `Expect: 100-continue` header, and I get a successful 200/OK response from the endpoint you supplied.

I see in your project.json that you're requesting netcoreapp1.1.  Is it possible that version isn't available in the AWS Lambda environment and that you're actually running on netcoreapp1.0?  https://github.com/dotnet/corefx/pull/10052 was in 1.1 and fixed the handling of Expect to respect this ExpectContinue setting, so I'd expect that to work for you as well if you're actually running on 1.1.

**Comment by karelz** [comment]
Does not seem to be 2.0-blocking issue based on the info above.

**Comment by stephentoub** [comment]
@JonCubed, could you try outputting the results of:
```
Console.WriteLine(FileVersionInfo.GetVersionInfo(typeof(object).GetTypeInfo().Assembly.Location));
Console.WriteLine(FileVersionInfo.GetVersionInfo(typeof(HttpClient).GetTypeInfo().Assembly.Location));
```
(or something other than Console.WriteLine, anything to see the results)
from inside of AWS Lambda?  I'm curious to see if maybe you're getting reverted to .NET Core 1.0 instead of 1.1, and the output you get should be telling.

The output will be something like:
```
stoub@SToubUbuntu1610:~/tmpapp$ dotnet run
File:             /usr/share/dotnet/shared/Microsoft.NETCore.App/1.1.1/System.Private.CoreLib.ni.dll
InternalName:     System.Private.CoreLib.ni.dll
OriginalFilename: System.Private.CoreLib.ni.dll
FileVersion:      4.6.25009.03
FileDescription:  System.Private.CoreLib
Product:          Microsoft/x00ae .NET Framework
ProductVersion:   4.6.25009.03. Commit Hash: 1735a1d453677717e68803da6a85284d15dca891
Debug:            False
Patched:          False
PreRelease:       False
PrivateBuild:     False
SpecialBuild:     False
Language:         Language Neutral

File:             /usr/share/dotnet/shared/Microsoft.NETCore.App/1.1.1/System.Net.Http.dll
InternalName:     System.Net.Http.dll
OriginalFilename: System.Net.Http.dll
FileVersion:      4.6.24705.01
FileDescription:  System.Net.Http
Product:          Microsoft® .NET Framework
ProductVersion:   4.6.24705.01. Commit Hash: 4d1af962ca0fede10beb01d197367c2f90e92c97
Debug:            False
Patched:          False
PreRelease:       False
PrivateBuild:     False
SpecialBuild:     False
Language:         Language Neutral
```
and those commit hashes in particular will tell us exactly what code is being used.

**Comment by JonCubed** [comment]
@stephentoub thanks for looking into this. Here is the output you requested

```
{
    "Comments": "System.Private.CoreLib",
    "CompanyName": "Microsoft Corporation",
    "FileBuildPart": 24326,
    "FileDescription": "System.Private.CoreLib",
    "FileMajorPart": 4,
    "FileMinorPart": 6,
    "FileName": "/var/lang/System.Private.CoreLib.ni.dll",
    "FilePrivatePart": 0,
    "FileVersion": "4.6.24326.0",
    "InternalName": "System.Private.CoreLib.ni.dll",
    "IsDebug": false,
    "IsPatched": false,
    "IsPrivateBuild": false,
    "IsPreRelease": false,
    "IsSpecialBuild": false,
    "Language": "Language Neutral",
    "LegalCopyright": "/x00a9 Microsoft Corporation.  All rights reserved.",
    "LegalTrademarks": "",
    "OriginalFilename": "System.Private.CoreLib.ni.dll",
    "PrivateBuild": "",
    "ProductBuildPart": 0,
    "ProductMajorPart": 0,
    "ProductMinorPart": 0,
    "ProductName": "Microsoft/x00ae .NET Framework",
    "ProductPrivatePart": 0,
    "ProductVersion": "4.6.24326.0 built by: p4admin-pb-worker-prod-i24xl-2c-al-e4d5a3fe.us-west-2.amazon.com. Commit Hash: N/A",
    "SpecialBuild": ""
}

{
    "Comments": "System.Net.Http",
    "CompanyName": "Microsoft Corporation",
    "FileBuildPart": 25214,
    "FileDescription": "System.Net.Http",
    "FileMajorPart": 1,
    "FileMinorPart": 0,
    "FileName": "/var/lang/System.Net.Http.dll",
    "FilePrivatePart": 0,
    "FileVersion": "1.0.25214.0",
    "InternalName": "System.Net.Http.dll",
    "IsDebug": false,
    "IsPatched": false,
    "IsPrivateBuild": false,
    "IsPreRelease": false,
    "IsSpecialBuild": false,
    "Language": "Language Neutral",
    "LegalCopyright": "/x00a9 Microsoft Corporation.  All rights reserved.",
    "LegalTrademarks": "",
    "OriginalFilename": "System.Net.Http.dll",
    "PrivateBuild": "",
    "ProductBuildPart": 0,
    "ProductMajorPart": 0,
    "ProductMinorPart": 0,
    "ProductName": "Microsoft/x00ae .NET Framework",
    "ProductPrivatePart": 0,
    "ProductVersion": "1.0.25214.0. Commit Hash: a147d50ba43b3bfb4534ac476cee6662616ab4b1",
    "SpecialBuild": ""
}
```

If I am understanding the product version correctly for `httpClient` it does look like it is using 1.0 version.

**Comment by stephentoub** [comment]
> thanks for looking into this

Happy to help.

> it does look like it is using 1.0 version

Yup.  That explains why you're seeing what you're seeing; it doesn't have the 1.1 fix around actually respecting ExpectContinue.

>  If I was running on a lower runtime shouldn't I be seeing an error because of the mismatch?

@eerhardt, do you know how someone could have a project.json like @JonCubed has specifying the 1.1 framework but still end up running on a 1.0 framework?

> I'll also check with aws support about this.

I think that's your best bet.  The underlying issue in corefx looks like it was fixed for 1.1, so I'm going to close this issue out.  If you feel it hasn't been fully addressed on the corefx side of things, please feel free to re-open. Thanks!

**Comment by eerhardt** [comment]
> @eerhardt, do you know how someone could have a project.json like @JonCubed has specifying the 1.1 framework but still end up running on a 1.0 framework?

The thing that controls it at runtime is the `runtimeconfig.json` file for your app.  Normally when you have this in your `project.json`:

```JSON
"dependencies": {
        "Microsoft.NETCore.App": {
            "type": "platform",
            "version": "1.1.0"
        },
```

that turns into:

```JSON
{
  "runtimeOptions": {
    "framework": {
      "name": "Microsoft.NETCore.App",
      "version": "1.1.0"
    }
  }
}
```

in your runtimeconfig.json file.  This tells the `dotnet` host to use the `1.1.0` or higher version of the shared framework to run your application.

I'm not familiar with Amazon Lambda, but if your `project.json` says `'type': 'platform', 'version': '1.1.0'` and your app is running on `1.0`, then my assumption is they are either re-writing your runtimeconfig.json, or not using yours and instead using their own.  (Or possibly even using their own host instead of the `dotnet` host.)
